### ISSUE TITLE
Something about netCDF-4.5.0-rc1 causes pio_decomp_frame_tests to hang

### ISSUE THREAD

**Issue by edhartnett** [NCAR/ParallelIO#1070]
I tried the 1.5.0-rc1 release of netCDF today. Unfortunately pio_decomp_frame_tests hangs. Works fine with netcdf-c-1.4.1.1.

Doh!

My approach will be to construct a C test that duplicates the problem. Jayesh, I believe this is your test - if you want to take a look, any assistance would be most appreciated.

**Comment by jayeshkrishna** [comment]
Could be a NetCDF bug (I have had issues in the past with newer versions or "different combinations of netcdf c+fortran+cxx" not working for me).
I will try the latest netcdf lib and see if I can recreate it. Meanwhile, you can go ahead and debug the issue.

**Comment by dqwu** [comment]
@jayeshkrishna 
I built netcdf-4.5.0-rc1 on anlworkstation, and I could not reproduce this issue so far:
85/92 Test #85: pio_decomp_frame_tests .............   Passed   14.98 sec
PS, netcdf-4.5.0-rc1 was built with hdf5-1.8.16

**Comment by edhartnett** [comment]
I am using hdf5-1.10.1. Maybe that's part of the problem.

I just updated to the latest version of all libraries on my main development machine...

**Comment by dqwu** [comment]
On anlworkstation, pio_decomp_frame_tests hangs if PIO is built with hdf5-1.10.0 (or higher versions) and netcdf-c-4.5.0-rc1.

Note:
hdf5-1.8.19 and netcdf-c-4.5.0-rc1 works.
hdf5-1.10.0/hdf5-1.10.1 and netcdf-c-4.4.1.1 also works.

It will be helpful to have a simplified C or Fortran test to reproduce this issue.

**Comment by edhartnett** [comment]
BTW this problem still exists using the netCDF HEAD (from yesterday). Many other memory problems were fixed, so it may be worthwhile for you guys to get a copy of the current netCDF code base, instead of using 4.5.0-rc1.

**Comment by dqwu** [comment]
The hanging issue of pio_decomp_frame_tests is caused by a recent enhancement added to netcdf-c-4.5.0-rc1:
_[Enhancement] Added support for HDF5 collective metadata operations when available. Patch submitted by Greg Sjaardema, see Pull request 335 for more information._

The detailed code changes can be viewed at https://github.com/Unidata/netcdf-c/pull/335

If HDF5 library has collective metadata APIs (HDF5 >= 1.10.0), HDF5_HAS_COLL_METADATA_OPS macro will be defined for netcdf-c-4.5.0-rc1. I commented out the following new code in libsrc4/nc4file.c and rebuilt NetCDF, pio_decomp_frame_tests could pass.
```
#ifdef HDF5_HAS_COLL_METADATA_OPS
   H5Pset_all_coll_metadata_ops(fapl_id, 1 );
   H5Pset_coll_metadata_write(fapl_id, 1);
#endif

#ifdef HDF5_HAS_COLL_METADATA_OPS
   H5Pset_all_coll_metadata_ops(fapl_id, 1 );
#endif
```
It is still not clear yet whether the fix would be on PIO side or netcdf-c side.

**Comment by edhartnett** [comment]
Good work finding this.

I don't think this is a bug in netCDF. Collective metadata calls are a
performance feature and we are in favor of performance, so I think we want
to support it.

Since only this one test fails, it seems like there is something about the
test that is not working like all the other tests. So we should change the
test so that it works, then continue CI including the latest versions of
HDF5 and netCDF-4, to ensure we don't inadvertently introduce this bug in
other places.

(And I would be very interested in hearing what this bug is all about. Just
making metadata calls collective should work, because, for non-async, all
code should be calling the HDF5 metadata functions in the same order.
Perhaps a barrier is needed somewhere in the test?)

Ed

**Comment by edhartnett** [comment]
I can confirm that it is the calls to HDF5 for collective metadata access which cause the problem. This happens in 7c68a255c9cf2723633a9082eeb966f955cc271f in git.

(Somewhat confusingly, for me, the change is not shown from the commit page. I had to clone the library before and after the commit and diff the directories to see the differences. I thought that's what github was showing me, but apparently not.)

**Comment by edhartnett** [comment]
I have been working this from the netCDF end. I have added pretty exhaustive testing to netCDF for parallel builds. (See https://github.com/Unidata/netcdf-c/pull/769).

I cannot find anything wrong with netCDF.

But I do have a concern. When reading the documentation of H5Pset_all_coll_metadata_ops() (See https://support.hdfgroup.org/HDF5/doc/RM/RM_H5P.html#Property-SetAllCollMetadataOps) I see this tantalizing clue:

> As noted above, corruption will be introduced into the metadata cache and HDF5 Library behavior will be undefined when both of the following conditions exist:
> A file is created or opened with a file access property list in which the collective metadata I/O property is set to TRUE.
> Any function is called that triggers an independent metadata read while the file remains open with that file access property list.

The default access mode is independent. So it seems that the current code will cause problems.

But when I try to create these conditions in a test, everything seems to work just fine.

I will open an issue on the netCDF project concerning this potential bug.

**Comment by edhartnett** [comment]
I have created a new C test, test_darray_4d.c which attempts to reproduce the actions of the Fortran test.

It is hanging at the same place, but I'm still investigating to see if I've reproduced the problem or just have a bug in my test code. ;-)

**Comment by jedwards4b** [comment]
@edhartnett I've just run tests on cheyenne and they all pass.  
```
100% tests passed, 0 tests failed out of 96

Total Test time (real) = 322.64 sec
jedwards@r1i1n21:~/sandboxes/ParallelIO/bld> module list

Currently Loaded Modules:
  1) ncarenv/1.2           4) mpt/2.15f       7) pylint/1.7.2
  2) intel/17.0.1          5) python/2.7.13   8) netcdf-mpi/4.5.0
  3) ncarcompilers/0.4.1   6) wrapt/1.10.10   9) pnetcdf/1.8.0

```

**Comment by edhartnett** [comment]
Try netcdf 4.6.1.

**Comment by edhartnett** [comment]
I am going to close this issue. I believe this issue was resolved by HDF5 1.10.4.
